### DISCUSSION TITLE
TileSet Custom Data Layer - Display Resource Path

### DISCUSSION THREAD

**Discussion by Jeroto** [godotengine/godot-proposals#8521] [Editor]
When applying `.tres` files to a custom data layer, instead of displaying either the file name or resource path, it displays a resource ID which is difficult to keep track of.
![image](https://github.com/godotengine/godot-proposals/assets/48961632/c20f4c86-7fe2-4840-b442-5ac7784ff89e)

It would be nice if there was a way to make the data more easily recognizable to the user, so they know what they've set on each tile. Perhaps there could be a type for the custom data layer which is specifically set up for resource files to handle this case?

Unless I'm missing something already implemented that does this, that is.

**Comment by Calinou** [comment]
> It would be nice if there was a way to make the data more easily recognizable to the user, so they know what they've set on each tile. Perhaps there could be a type for the custom data layer which is specifically set up for resource files to handle this case?

All Resources have a Name property that can be set by the user for identification purposes, but it's not used in the editor as often as it could be.

I made a quick branch for testing, which seems to work correctly: https://github.com/Calinou/godot/tree/tileset-editor-display-resource-name

![Screenshot_20231127_154520](https://github.com/godotengine/godot-proposals/assets/180032/f6587b51-ecda-4a14-89dc-b8f4975d06dc)

The resource on the left is named, the one on the right isn't. Maybe there should be a way to disambiguate unique resources that have the same name.

**Reply by Jeroto** [reply]
Just compiled and tried out the branch, and that's really nice!
![image](https://github.com/godotengine/godot-proposals/assets/48961632/ab6fd05c-16a3-4c52-864f-47ee22c83785)

Honestly I hadn't noticed or used the `Name` property for resources before so I'll definitely be using that more in the future as well.
Being able to tell resources with the same name apart would be nice, I'm not quite sure how to go about designing that in a meaningful way without using the file path though.
Even as-is I think it would be a nice addition to the engine, at least for what I was doing with it.

**Reply by Calinou** [reply]
> Being able to tell resources with the same name apart would be nice, I'm not quite sure how to go about designing that in a meaningful way without using the file path though.

We could keep displaying the resource ID on a new line when a resource is named, but the issue is that it's pretty long. There are a few ways it could be shortened:

- Only display the last 7 digits or so. This will shorten the displayed value a lot, but it won't be reversable anymore into an actual object ID for debugging and there's a (very) small risk of collisions.
- Display as hexadecimal, but this only shortens it by 2-3 characters on average. The hexadecimal representation of negative numbers (which are regularly used for resource object IDs) is also a bit uncommon, so it may surprise some people too.
- Display as an encoded string that uses all 62 uppercase/lowercase/number characters. This can result in much shorter IDs, but this would be a nonstandard hardcoding that is harder to reverse for debugging purposes (e.g. to compare with `get_instance_id()`'s output).
